### PR TITLE
Refactor MipLibTrick preprocessing pass.

### PR DESCRIPTION
**Pull request by mpreiner**

### CODE REVIEW

File: src/preprocessing/passes/miplib_trick.cpp [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+            Debug("miplib")
+                << "skipSubs: " << newAssertion[0] << " to " << newAssertion[1]
+                << " (threshold is " << options::arithMLTrickSubstitutions()
+                << ")" << endl;
+          }
+          newAssertion = Rewriter::rewrite(newAssertion);
+          Debug("miplib") << "  " << newAssertion << endl;
+
+          assertionsToPreprocess->push_back(newAssertion);
+          ProofManager::currentPM()->addDependence(newAssertion, Node::null());
```
Comment (4tXJ7f) [inline src/preprocessing/passes/miplib_trick.cpp]: Ditto

File: src/preprocessing/passes/miplib_trick.cpp [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+                  ss.str(),
+                  nm->integerType(),
+                  "a variable introduced due to scrubbing a miplib encoding",
+                  NodeManager::SKOLEM_EXACT_NAME);
+              Node geq = Rewriter::rewrite(nm->mkNode(kind::GEQ, newVar, zero));
+              Node leq = Rewriter::rewrite(nm->mkNode(kind::LEQ, newVar, one));
+
+              Node n = Rewriter::rewrite(geq.andNode(leq));
+              assertionsToPreprocess->push_back(n);
+              ProofManager::currentPM()->addDependence(n, Node::null());
```
Comment (4tXJ7f) [inline src/preprocessing/passes/miplib_trick.cpp]: This needs to be protected by an `if (PROOF_ON())` or in a `PROOF()` macro. If CVC4 is compiled without proof support [`currentPM()` will return a `nullptr`](https://github.com/CVC4/CVC4/blob/047e75b485ad16a729083c210ba4064943d2e7c5/src/smt/smt_engine_scope.cpp#L44).

Comment (mpreiner) [reply src/preprocessing/passes/miplib_trick.cpp]: Done

File: src/preprocessing/passes/miplib_trick.cpp (hunk 1 above)
Comment (mpreiner) [reply src/preprocessing/passes/miplib_trick.cpp]: Done

File: src/smt/smt_engine.cpp [hunk 3]
```diff
@@ -3806,9 +3388,8 @@ bool SmtEnginePrivate::simplifyAssertions()
         Trace("simplify") << "SmtEnginePrivate::simplify(): "
```
[the remark this answers is not in the document]
Comment (mpreiner) [reply src/smt/smt_engine.cpp]: Done

File: src/preprocessing/passes/miplib_trick.cpp [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+  {
+    d_boolVars.push_back(n);
+  }
+}
+
+void MipLibTrick::nmNotifyNewSkolem(TNode n,
+                                    const std::string& comment,
+                                    uint32_t flags)
+{
+  if (n.getType().isBoolean() && !options::incrementalSolving())
```
[the remark this answers is not in the document]
Comment (mpreiner) [reply src/preprocessing/passes/miplib_trick.cpp]: Done

File: src/preprocessing/preprocessing_pass_context.h [hunk 5]
```diff
@@ -25,25 +25,36 @@
...
 #include "smt/smt_engine.h"
 #include "smt/term_formula_removal.h"
+#include "theory/booleans/circuit_propagator.h"
 #include "theory/theory_engine.h"
 #include "util/resource_manager.h"
 
 namespace CVC4 {
 namespace preprocessing {
 
+using namespace theory;
```
[the remark this answers is not in the document]
Comment (mpreiner) [reply src/preprocessing/preprocessing_pass_context.h]: Oh how did that happen! :-) Done.

**Pull request merged**